### ISSUE TITLE
OpenSlides not usable - long wait times

### ISSUE THREAD

**Issue by topelrapha** [OpenSlides/openslides-docker-compose#7]
Is it your intention to have a long wait time for redis cache?

WAIT_TIMEOUT = 10000 in server/entrypoint

Our platform is not usable - but changing the value to 1000 - it works.

Thanks for your support!

**Comment by ManniLoe** [comment]
PR #8 points to the same problem

**Comment by FinnStutzenstein** [comment]
Hi,
this timeout should never be reached, if the usage is low (a handful of participants, no extensive voting, ...). It is high to compensate redis repliation times, if the setup is highly scaled used across multiple servers in a docker stack configuration. If this timeout is always reached it is an indication, that your redis setup is incorrect.

Can you give some details about the setup

- Docker compose or docker stack
- Did you just use the default config, or did you change something related the the points below:
- The value of `REDIS_REPLICAS` in the docker-*.yml
- The section of `redis-slave` in the docker-*.yml

I can see, that there is one case, where the default is misconfigured; But first I want to make sure, that this is the problem you have.

Regarding the PR: Lowering the timeout does not fix the problem, since every request is now 1 second delayed instead of 10, but there is still an unnecessary delay.

**Comment by topelrapha** [comment]
Thanks for your reply.

We are using docker compose and leave all configs default:

`REDIS_REPLICAS: 1 # must match redis-slave replicas below!`
` redis-slave:
    image: redis:alpine
    restart: always
    command: ["redis-server", "--save", "", "--slaveof", "redis", "6379"]
    depends_on:
      - redis
    networks:
      back:
        aliases:
          - rediscache-slave
  redis-channels:
    image: redis:alpine
    restart: always
    networks:
      back:
`

**Comment by FinnStutzenstein** [comment]
Ok, are you sure, that `REDIS_REPLICAS` is 1? Can you confirm this in the startup logs? There should be a log statement printing out this value. Maybe you can post the printed section (there are multiple redis-related configuration printouts).

If it is really 1, I'll test this tomorrow.

There were updates of this repo lately. Which commit are you using for the configuration?

**Comment by topelrapha** [comment]
Commit is: a46a5dd007d64a2964b1a3c647ecbefd42caaf6b

And part of the container log:
server_1          | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis_connection_pool [HBWf] CONNECTION_POOL_LIMIT=100
server_1          | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis [HBWf] Redis address redis://redis:6379/0
server_1          | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis [HBWf] Redis read only address redis://redis-slave:6379/0
server_1          | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis [HBWf] AMOUNT_REPLICAS=1
server_1          | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis [HBWf] WAIT_TIMEOUT=1000

**Comment by FinnStutzenstein** [comment]
Hi,

After setting `AMOUNT_REPLICAS` in the `docker-compose.yml` to 1 (the default is 3, which is wrong) I can start the setup as described in the README without this error. Having a wrong `AMOUNT_REPLICAS` leads to

```
openslides.utils.cache_providers [HjRx] WAIT reported 1 replicas of 3 requested after 10000 ms!
```

Do you have this statement in your logs, if you run into the timeout? (Maybe put it back to 10s for testing)

Otherwiese you can do two things:

1) Disable the read only redis: There is no nice way, you have to modify the `server/entrypoint` and remove this statement: https://github.com/OpenSlides/openslides-docker-compose/blob/master/server/entrypoint#L55 Without it, the code leading to the timeout is disabled. You can optionally remove the `redis-slave` from the `docker-compose.yml`. The timeout should not accur anymore. The redis slave is not strictly needed. As I said: If you have assemblies without using electronic voting or other highly demanding stuff, it should be sufficient.

2) Check, if the replication is working. Use `docker-compose exec redis redis-cli` and basic tedis commands to change something in `redis`. Meanwhile in another terminal exec into `redis-slave` and check, if changes are propagated correctly. Refer to https://redis.io/commands#generic. You can use `SET`, `GET`, `DEL` for a simple interaction with redis.

**Comment by jfwiebe** [comment]
Hi!
I've encountered the same problem and solved the problem by changing `AMOUNT_REPLICAS` to 1. The bigger problem during my debugging was that the old docker volumes are not deleted by `docker-compose down`, so the `AMOUNT_REPLICAS` in the database stayed at 3. To ensure that you are starting with fresh containers and volumes, perform `docker-compose down`, `docker-compose rm` and `docker volume prune` (attention, this will delete all volumes with no container linked to it, also non-openslides-related!). After that, you can rebuild everything and the change takes effect.

I think, that the [docker-compose.yml.example](https://github.com/OpenSlides/openslides-docker-compose/blob/master/docker-compose.yml.example) (and swarm, too) should be adjusted to be consistent (`AMOUNT_REPLICAS` to 1 and one redis-slave or `AMOUNT_REPLICAS` to 3 and three redis-slave containers). I think, I'd prefer the first option.

Greetings,
clarity

**Comment by FinnStutzenstein** [comment]
> The bigger problem during my debugging was that the old docker volumes are not deleted by docker-compose down, so the AMOUNT_REPLICAS in the still-existing database stayed at 3.

Yep, I struggled at this point, too.

>  `docker-compose rm` and `docker volume prune`

:+1: 

>  or `AMOUNT_REPLICAS` to 3 and three redis-slave containers

This does not work for docker compose when keeping it comparable to stack: There is no scaling for docker compose. The first suggested change is what I have refered to in my first comment ("I can see, that there is one case, where the default is misconfigured") and want to check, that this is not the reason for the reported bug.

> So from my understanding, lowering the WAIT_TIMEOUT as stated in #8 is only fighting symtoms here but not solving the problem.

This is what I'm trying to figure out.

The strange problem is, taht in the log line from @topelrapha it explicilty says

```
server_1 | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis [HBWf] AMOUNT_REPLICAS=1
```

So it is not the issue, that `AMOUNT_REPLICAS` has the wrong default.

**Comment by jfwiebe** [comment]
> > or `AMOUNT_REPLICAS` to 3 and three redis-slave containers
> 
> This does not work for docker compose when keeping it comparable to stack: There is no scaling for docker compose. The first suggested change is what I have referred to in my first comment ("I can see, that there is one case, where the default is misconfigured") and want to check, that this is not the reason for the reported bug.

Thanks for that explanation - I've understood the problem now. In the stack config, the `AMOUNT_REPLICAS` of three because the "redis-slave" will be spawned three times (attribute "[replica](https://docs.docker.com/compose/compose-file/#replicas)"); but in the compose config, `AMOUNT_REPLICAS` has to be 1 by default because there are no replicas configured. In that case, I would love to create a pull request to take the general issue into a separate topic and focus on @topelrapha's problem.

> The strange problem is, taht in the log line from @topelrapha it explicilty says
> 
> ```
> server_1 | [2020-06-25 03:57:51 +0200] [1] [INFO] openslides.utils.redis [HBWf] AMOUNT_REPLICAS=1
> ```

Got it now, sorry for my misunderstanding earlier. In that case, I can't give a better advice than ensuring that the volumes need to be fully reset (with the docker commands stated above) and the `AMOUNT_REPLICAS` is set to 1. In my setup (current debian stable, current docker/docker compose, current openslides), it works like a charm.

**Comment by FinnStutzenstein** [comment]
> In that case, I would love to create a pull request to take the general issue into a separate topic and focus on @topelrapha's problem.

This would be great!

> In my setup (current debian stable, current docker/docker compose, current openslides), it works like a charm.

Yes, I cannot reproduce this either.  @topelrapha Should check both suggestions above: Disabling the read-replicat (so the source code for the bud is disabled) and checking, if the replication/redis-cache works.

These are the only two things I can image that do not work to trigger the timeout with a correctly `AMOUNT_REPLICAS`

**Comment by gsiv** [comment]
>  but in the compose config, `AMOUNT_REPLICAS` has to be 1 by default because there are no replicas configured.

The default setting of 3 for the Compose setup was indeed a mistake and I've changed it back to 1.

> > or `AMOUNT_REPLICAS` to 3 and three redis-slave containers
> 
> This does not work for docker compose when keeping it comparable to stack: There is no scaling for docker compose.

This is not entirely correct.  Replica settings work with docker-compose in `--compatibility` mode.  It is non-standard behavior, however, so AMOUNT_REPLICAS should definitely still be set to 1 by default.

In case there is demand for it, I have added conditional scaling options to `docker-compose.yml.m4`, see https://github.com/OpenSlides/openslides-docker-compose/blob/921c0c33ae2c5c3ca488eef1aee8082837f498cc/docker-compose.yml.m4#L172-L173.
